File: disease_prediction.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <time.h>

#define MAX_PATIENTS 50
#define MAX_DISEASES 50
#define DNA_LENGTH 14

struct DiseaseRisk {
    char diseaseName[50];
    float percentage;
    char label[30];
};

struct Patient {
    char name[50];
    int age;
    char dob[15];
    char bloodGroup[5];
    float height;
    float weight;
    float mutation_pct;
    struct DiseaseRisk risks[MAX_DISEASES];
    int riskCount;
};

struct Disease {
    char name[50];
    char symptoms[100];
    char code[10];
};

struct RiskLevel {
    char label[30];
    int min;
    int max;
    char precautions[100];
};

struct Patient patients[MAX_PATIENTS];
int patientCount = 0;

struct Disease diseases[MAX_DISEASES];
int diseaseCount = 0;

const char NORMAL_DNA[] = "ATGCGTACGTTAGC";

struct RiskLevel riskChart[] = {
    {"No Risk", 0, 10, "Maintain healthy lifestyle."},
    {"Low Risk", 11, 30, "Regular checkups, balanced diet."},
    {"Moderate Risk", 31, 60, "Consult doctor, monitor symptoms, specific diet."},
    {"High Risk", 61, 100, "Immediate medical attention, medication."}
};
int numRiskLevels = 4;

void patientProfile() {
    if (patientCount >= MAX_PATIENTS) {
        printf("Error: Patient database is full!\n");
        return;
    }
    printf("\n--- Add Patient Profile ---\n");
    printf("Enter Name: ");
    scanf(" %[^\n]", patients[patientCount].name);
    printf("Enter Age: ");
    scanf("%d", &patients[patientCount].age);
    printf("Enter Date of birth (DD/MM/YYYY): ");
    scanf(" %s", patients[patientCount].dob);
    printf("Enter Blood group (A+/B+/AB+/O+/A-/B-/AB-/O-): ");
    scanf(" %s", patients[patientCount].bloodGroup);
    printf("Enter Height (cm): ");
    scanf("%f", &patients[patientCount].height);
    printf("Enter Weight (kg): ");
    scanf("%f", &patients[patientCount].weight);
    
    patients[patientCount].mutation_pct = 0.0;
    patients[patientCount].riskCount = 0;
    
    patientCount++;
    printf("Patient added successfully!\n");
}

void diseaseProfile() {
    if (diseaseCount >= MAX_DISEASES) {
        printf("Error: Disease database is full!\n");
        return;
    }
    printf("\n--- Add Disease Profile ---\n");
    printf("Enter Disease Name: ");
    scanf(" %[^\n]", diseases[diseaseCount].name);
    printf("Enter Symptoms: ");
    scanf(" %[^\n]", diseases[diseaseCount].symptoms);
    printf("Enter Reference Code: ");
    scanf(" %s", diseases[diseaseCount].code);
    
    diseaseCount++;
    printf("Disease added successfully!\n");
}

void referenceTables() {
    int choice;
    printf("\n--- Reference Tables ---\n");
    printf("1) Disease Chart\n");
    printf("2) Risk Chart\n");
    printf("Enter a number: ");
    
    if (scanf("%d", &choice) != 1) {
        while(getchar() != '\n');
        printf("Error: Invalid Input\n");
        return;
    }
    
    if (choice == 1) {
        printf("\n--- Disease Chart ---\n");
        if (diseaseCount == 0) {
            printf("No diseases in the chart yet.\n");
        } else {
            for (int i = 0; i < diseaseCount; i++) {
                printf("Name: %-15s | Code: %-5s | Symptoms: %s\n", diseases[i].name, diseases[i].code, diseases[i].symptoms);
            }
        }
    } else if (choice == 2) {
        printf("\n--- Risk Chart ---\n");
        for (int i = 0; i < numRiskLevels; i++) {
            printf("Label: %-15s | Range: %2d-%-3d%% | Precautions: %s\n", riskChart[i].label, riskChart[i].min, riskChart[i].max, riskChart[i].precautions);
        }
    } else {
        printf("Error: Invalid Input\n");
    }
}

void patientRecords() {
    printf("\n--- Patient Records ---\n");
    if (patientCount == 0) {
        printf("No patient entries found.\n");
        return;
    }
    for (int i = 0; i < patientCount; i++) {
        printf("\nName: %-15s | Age: %-3d | Blood: %-3s | Height: %-5.1f | Weight: %-5.1f\n", 
               patients[i].name, patients[i].age, patients[i].bloodGroup, patients[i].height, patients[i].weight);
        
        if (patients[i].riskCount > 0) {
            printf("Assigned Risk Labels:\n");
            for (int j = 0; j < patients[i].riskCount; j++) {
                printf("  - %s: %s (%.2f%%)\n", patients[i].risks[j].diseaseName, patients[i].risks[j].label, patients[i].risks[j].percentage);
            }
        } else {
            printf("Risks: No diseases checked yet.\n");
        }
    }
}

int selectPatient() {
    if (patientCount == 0) {
        printf("No patients available.\n");
        return -1;
    }
    for (int i = 0; i < patientCount; i++) {
        printf("%d) %s\n", i + 1, patients[i].name);
    }
    int choice;
    printf("Select patient number: ");
    if (scanf("%d", &choice) != 1 || choice < 1 || choice > patientCount) {
        while(getchar() != '\n'); 
        printf("Error: Invalid Input\n");
        return -1;
    }
    return choice - 1;
}

int selectDisease() {
    if (diseaseCount == 0) {
        printf("No diseases available in the chart.\n");
        return -1;
    }
    for (int i = 0; i < diseaseCount; i++) {
        printf("%d) %s\n", i + 1, diseases[i].name);
    }
    int choice;
    printf("Select disease number: ");
    if (scanf("%d", &choice) != 1 || choice < 1 || choice > diseaseCount) {
        while(getchar() != '\n');
        printf("Error: Invalid Input\n");
        return -1;
    }
    return choice - 1;
}

void mutationAnalysis() {
    printf("\n--- Mutation Analysis ---\n");
    int pIdx = selectPatient();
    if (pIdx == -1) return;
    
    printf("\nReference DNA: %s\n", NORMAL_DNA);
    printf("Enter DNA strand code for patient: ");
    char inputDNA[50];
    scanf(" %s", inputDNA);
    
    if (strlen(inputDNA) != DNA_LENGTH) {
        printf("Error: Entered DNA strand must be exactly %d characters long.\n", DNA_LENGTH);
        return;
    }
    
    int diffs = 0;
    for (int i = 0; i < DNA_LENGTH; i++) {
        if (inputDNA[i] != NORMAL_DNA[i]) {
            diffs++;
        }
    }
    
    float pct = ((float)diffs / DNA_LENGTH) * 100.0f;
    patients[pIdx].mutation_pct = pct;
    
    printf("\nLevel of mutation present: %.2f%%\n", pct);
    
    if (diffs > 0) {
        printf("Specific Mutation Details (Codon Analysis):\n");
        for (int i = 0; i < DNA_LENGTH; i += 3) {
            char cNorm[4] = {0}, cIn[4] = {0};
            strncpy(cNorm, &NORMAL_DNA[i], 3);
            strncpy(cIn, &inputDNA[i], 3);
            
            if (strcmp(cNorm, cIn) != 0) {
                printf("  - Block %d: Mutated to '%s', normally should be '%s'\n", (i/3)+1, cIn, cNorm);
            }
        }
    } else {
        printf("No mutations detected.\n");
    }
}

void diseaseRiskPrediction() {
    printf("\n--- Disease Risk Prediction ---\n");
    int pIdx = selectPatient();
    if (pIdx == -1) return;
    
    int dIdx = selectDisease();
    if (dIdx == -1) return;
    
    float baseRisk = patients[pIdx].mutation_pct * 1.5;
    float environmental = (rand() % 30) + 10.0;
    float totalRisk = baseRisk + environmental;
    if (totalRisk > 100.0) totalRisk = 100.0;
    
    char assignedLabel[30] = "No Risk";
    for (int i = 0; i < numRiskLevels; i++) {
        if (totalRisk >= riskChart[i].min && totalRisk <= riskChart[i].max) {
            strcpy(assignedLabel, riskChart[i].label);
            break;
        }
    }
    
    int rIdx = patients[pIdx].riskCount;
    if (rIdx < MAX_DISEASES) {
        strcpy(patients[pIdx].risks[rIdx].diseaseName, diseases[dIdx].name);
        patients[pIdx].risks[rIdx].percentage = totalRisk;
        strcpy(patients[pIdx].risks[rIdx].label, assignedLabel);
        patients[pIdx].riskCount++;
    }
    
    printf("\n[Analysis Complete] Patient: %s | Disease: %s\n", patients[pIdx].name, diseases[dIdx].name);
    printf("Calculated Risk Level: %.2f%%\n", totalRisk);
    printf("Assigned Risk Label: %s\n", assignedLabel);
}

void drawAsciiDemographicsChart(int sus, int inf, int rec) {
    int total = sus + inf + rec;
    if (total == 0) return;
    
    int barWidth = 40;
    int susLen = (int)(((float)sus / total) * barWidth);
    int infLen = (int)(((float)inf / total) * barWidth);
    int recLen = barWidth - susLen - infLen;
    
    printf("\n--- Demographic Chart (Terminal ASCII Block Representation) ---\n");
    printf("Susceptible | ");
    for(int i=0; i<susLen; i++) printf("\xB1");
    printf(" [%.1f%%]\n", ((float)sus/total)*100);
    
    printf("Infected    | ");
    for(int i=0; i<infLen; i++) printf("\xDB");
    printf(" [%.1f%%]\n", ((float)inf/total)*100);
    
    printf("Recovered   | ");
    for(int i=0; i<recLen; i++) printf("\xB0");
    printf(" [%.1f%%]\n", ((float)rec/total)*100);
    printf("-------------------------------------------------------------\n");
}

void pandemicSpreadSimulation() {
    printf("\n--- Pandemic Spread Simulation ---\n");
    int dIdx = selectDisease();
    if (dIdx == -1) return;
    
    int days = (rand() % 90) + 10;
    int totalPop = 100000;
    int sus = (rand() % 50000) + 20000;
    int rec = (rand() % 20000) + 5000;
    int inf = totalPop - sus - rec;
    
    printf("\nDisease name: %s\n", diseases[dIdx].name);
    printf("Number of days since onset: %d\n", days);
    printf("Number of individuals susceptible: %d\n", sus);
    printf("Number of individuals infected: %d\n", inf);
    printf("Number of people recovered: %d\n", rec);
    
    drawAsciiDemographicsChart(sus, inf, rec);
}

int main() {
    int choice;
    srand(time(NULL));
    
    while(1) {
        printf("\n=============================================\n");
        printf("1) Patient Profile\n");
        printf("2) Disease Profile\n");
        printf("3) Reference Tables\n");
        printf("4) Patient Records\n");
        printf("5) Mutation Analysis\n");
        printf("6) Disease Risk Prediction\n");
        printf("7) Pandemic Spread Simulation\n");
        printf("8) Exit\n");
        printf("=============================================\n");
        printf("Enter a number: ");
        
        if (scanf("%d", &choice) != 1) {
            while(getchar() != '\n'); 
            printf("Error: Invalid Input\n");
            continue;
        }
        
        switch(choice) {
            case 1: patientProfile(); break;
            case 2: diseaseProfile(); break;
            case 3: referenceTables(); break;
            case 4: patientRecords(); break;
            case 5: mutationAnalysis(); break;
            case 6: diseaseRiskPrediction(); break;
            case 7: pandemicSpreadSimulation(); break;
            case 8: printf("Exiting...\n"); return 0;
            default: printf("Error: Invalid Input\n");
        }
    }
    return 0;
}

File: disease_prediction.py
import numpy as np
import matplotlib.pyplot as plt

patients = []
diseases = []
MAX_PATIENTS = 50

NORMAL_DNA = "ATGCGTACGTTAGC"

risk_chart = [
    {"label": "No Risk", "min": 0, "max": 10, "precautions": "Maintain healthy lifestyle."},
    {"label": "Low Risk", "min": 11, "max": 30, "precautions": "Regular checkups, balanced diet."},
    {"label": "Moderate Risk", "min": 31, "max": 60, "precautions": "Consult doctor, monitor symptoms, specific diet."},
    {"label": "High Risk", "min": 61, "max": 100, "precautions": "Immediate medical attention, medication, strict isolation."}
]


def display_menu():
    print("\n=============================================")
    print("1) Patient Profile")
    print("2) Disease Profile")
    print("3) Reference Tables")
    print("4) Patient Records")
    print("5) Mutation Analysis")
    print("6) Disease Risk Prediction")
    print("7) Pandemic Spread Simulation")
    print("8) Exit")
    print("=============================================")


def patient_profile():
    if len(patients) >= MAX_PATIENTS:
        print("Error: Patient database is full!")
        return

    print("\n--- Add Patient Profile ---")
    name = input("Enter Name: ")
    age = input("Enter Age: ")
    dob = input("Enter Date of Birth (DD/MM/YYYY): ")
    blood_group = input("Enter Blood group (A+/B+/AB+/O+/A-/B-/AB-/O-): ")
    height = input("Enter Height (cm): ")
    weight = input("Enter Weight (kg): ")

    patients.append({
        "name": name,
        "age": age,
        "dob": dob,
        "blood_group": blood_group,
        "height": height,
        "weight": weight,
        "mutation_pct": 0.0,
        "risks": {}
    })
    print("Patient added successfully!")


def disease_profile():
    print("\n--- Add Disease Profile ---")
    name = input("Enter Disease Name: ")
    symptoms = input("Enter Symptoms: ")
    ref_code = input("Enter Reference Code: ")

    diseases.append({
        "name": name,
        "symptoms": symptoms,
        "code": ref_code
    })
    print("Disease added successfully!")


def reference_tables():
    while True:
        print("\n--- Reference Tables ---")
        print("1) Disease Chart")
        print("2) Risk Chart")
        choice = input("Enter a number: ")

        if choice == '1':
            print("\n--- Disease Chart ---")
            if not diseases:
                print("No diseases in the chart yet.")
            else:
                for d in diseases:
                    print(f"Name: {d['name']:<15} | Code: {d['code']:<5} | Symptoms: {d['symptoms']}")
            break
        elif choice == '2':
            print("\n--- Risk Chart ---")
            for r in risk_chart:
                print(f"Label: {r['label']:<15} | Range: {r['min']:>2}-{r['max']:<3}% | Precautions: {r['precautions']}")
            break
        else:
            print("Error: Invalid Input. Please enter 1 or 2.")


def patient_records():
    print("\n--- Patient Records ---")
    if not patients:
        print("No patient entries found.")
        return

    for p in patients:
        print(f"\nName: {p['name']:<15} | Age: {p['age']:<3} | Blood: {p['blood_group']:<3} | Height: {p['height']:<5} | Weight: {p['weight']:<5}")
        if p['risks']:
            print("Assigned Risk Labels:")
            for d_name, risk_data in p['risks'].items():
                print(f"  - {d_name}: {risk_data['label']} ({risk_data['percentage']:.2f}%)")
        else:
            print("Risks: No diseases checked yet.")


def get_patient():
    if not patients:
        print("No patients available.")
        return None
    for idx, p in enumerate(patients):
        print(f"{idx + 1}) {p['name']}")
    try:
        choice = int(input("Select patient number: ")) - 1
        if 0 <= choice < len(patients):
            return patients[choice]
        else:
            print("Invalid selection.")
            return None
    except ValueError:
        print("Error: Invalid Input")
        return None


def get_disease():
    if not diseases:
        print("No diseases available in the chart.")
        return None
    for idx, d in enumerate(diseases):
        print(f"{idx + 1}) {d['name']}")
    try:
        choice = int(input("Select disease number: ")) - 1
        if 0 <= choice < len(diseases):
            return diseases[choice]
        else:
            print("Invalid selection.")
            return None
    except ValueError:
        print("Error: Invalid Input")
        return None


def mutation_analysis():
    print("\n--- Mutation Analysis ---")
    p = get_patient()
    if not p: return

    print(f"\nReference DNA: {NORMAL_DNA}")
    input_dna = input("Enter DNA strand code for patient: ").upper()
    
    if len(input_dna) != len(NORMAL_DNA):
        print(f"Error: Entered DNA strand must be {len(NORMAL_DNA)} characters long.")
        return

    arr_normal = np.array(list(NORMAL_DNA))
    arr_input = np.array(list(input_dna))

    mutations = (arr_normal != arr_input)
    
    mutation_count = np.sum(mutations)
    mutation_pct = (mutation_count / len(NORMAL_DNA)) * 100
    p['mutation_pct'] = mutation_pct

    print(f"\nLevel of mutation present: {mutation_pct:.2f}%")

    if mutation_count > 0:
        print("Specific Mutation Details (Codon Analysis):")
        for i in range(0, len(NORMAL_DNA), 3):
            codon_normal = "".join(arr_normal[i:i+3])
            codon_input = "".join(arr_input[i:i+3])
            
            if codon_normal != codon_input:
                print(f"  - Codon Block {i//3 + 1}: Mutated to '{codon_input}', normally should be '{codon_normal}'")
    else:
        print("No mutations present.")


def disease_risk_prediction():
    print("\n--- Disease Risk Prediction ---")
    p = get_patient()
    if not p: return

    d = get_disease()
    if not d: return

    base_mutation_risk = p.get('mutation_pct', 0.0)

    susceptibility_factor = np.random.uniform(10.0, 40.0)
    total_risk = min(100.0, (base_mutation_risk * 1.5) + susceptibility_factor)

    assigned_label = "No Risk"
    for r in risk_chart:
        if r['min'] <= total_risk <= r['max']:
            assigned_label = r['label']
            break
    
    p['risks'][d['name']] = {
        'percentage': total_risk,
        'label': assigned_label
    }

    print(f"\n[Analysis Complete] Patient: {p['name']} | Disease: {d['name']}")
    print(f"Calculated Risk Level: {total_risk:.2f}%")
    print(f"Assigned Risk Label: {assigned_label}")


def pandemic_spread_simulation():
    print("\n--- Pandemic Spread Simulation ---")
    d = get_disease()
    if not d: return

    days_since_onset = int(np.random.randint(10, 100))
    total_population = 100000

    susceptible = int(np.random.randint(20000, 70000))
    recovered = int(np.random.randint(5000, 25000))
    infected = total_population - susceptible - recovered

    print(f"\nDisease name: {d['name']}")
    print(f"Number of days since onset: {days_since_onset}")
    print(f"Number of individuals susceptible: {susceptible}")
    print(f"Number of individuals infected: {infected}")
    print(f"Number of people recovered: {recovered}")

    print("Generating simulation pie chart display...")
    
    labels = ['Susceptible', 'Infected', 'Recovered']
    sizes = [susceptible, infected, recovered]
    colors = ['#8fc3a3', '#d9534f', '#5bc0de']
    
    plt.figure(figsize=(7, 7))
    plt.pie(sizes, labels=labels, colors=colors, autopct='%1.1f%%', startangle=140, explode=(0.05, 0.05, 0.05))
    plt.title(f"Pandemic Spread Simulation - {d['name']}")
    plt.axis('equal')
    
    plt.show()


def main():
    while True:
        display_menu()
        choice = input("Enter a number: ")

        if choice == '1':
            patient_profile()
        elif choice == '2':
            disease_profile()
        elif choice == '3':
            reference_tables()
        elif choice == '4':
            patient_records()
        elif choice == '5':
            mutation_analysis()
        elif choice == '6':
            disease_risk_prediction()
        elif choice == '7':
            pandemic_spread_simulation()
        elif choice == '8':
            print("Exiting...")
            break
        else:
            print("Error: Invalid Input")


if __name__ == "__main__":
    main()